gflownet/envs/tree.py

```diff
@@ -0,0 +1,1079 @@
+        if torch.any(mask_discrete):
+            policy_outputs_discrete = policy_outputs[
+                mask_discrete, : self._index_continuous_policy_output
+            ]
+            actions_discrete, logprobs_discrete = super().sample_actions(
+                policy_outputs_discrete,
+                sampling_method,
+                mask_invalid_actions[
+                    mask_discrete, : self._index_continuous_policy_output
+                ],
+                temperature_logits,
+            )
+            logprobs[mask_discrete] = logprobs_discrete
+        if torch.all(mask_discrete):
+            return actions_discrete, logprobs
+        # Continuous actions
+        mask_cont = torch.logical_not(mask_discrete)
+        n_cont = mask_cont.sum()
+        policy_outputs_cont = policy_outputs[
+            mask_cont, self._index_continuous_policy_output :
+        ]
+        if sampling_method == "uniform":
+            distr_threshold = Uniform(
+                torch.zeros(n_cont),
+                torch.ones(n_cont),
+            )
+        elif sampling_method == "policy":
+            mix_logits = policy_outputs_cont[:, 0::3]
+            mix = Categorical(logits=mix_logits)
+            alphas = policy_outputs_cont[:, 1::3]
+            alphas = self.beta_params_max * torch.sigmoid(alphas) + self.beta_params_min
+            betas = policy_outputs_cont[:, 2::3]
+            betas = self.beta_params_max * torch.sigmoid(betas) + self.beta_params_min
+            beta_distr = Beta(alphas, betas)
+            distr_threshold = MixtureSameFamily(mix, beta_distr)
+        thresholds = distr_threshold.sample()
+        # Log probs
+        logprobs[mask_cont] = distr_threshold.log_prob(thresholds)
+        # Build actions
+        actions_cont = [(ActionType.PICK_THRESHOLD, th.item()) for th in thresholds]
+        actions = []
+        for is_discrete in mask_discrete:
+            if is_discrete:
+                actions.append(actions_discrete.pop(0))
+            else:
+                actions.append(actions_cont.pop(0))
+        return actions, logprobs
+
+    def get_logprobs(
+        self,
+        policy_outputs: TensorType["n_states", "policy_output_dim"],
+        is_forward: bool,
+        actions: TensorType["n_states", "n_dim"],
+        states_target: TensorType["n_states", "policy_input_dim"],
+        mask_invalid_actions: TensorType["n_states", "1"] = None,
+    ) -> TensorType["batch_size"]:
+        """
+        Computes log probabilities of actions given policy outputs and actions.
+        """
+        n_states = policy_outputs.shape[0]
+        if states_target is None:
+            states_target = torch.empty(
+                (n_states, self.policy_input_dim), device=self.device
+            )
+        logprobs = torch.zeros(n_states, device=self.device, dtype=self.float)
+        # Discrete actions
+        mask_discrete = mask_invalid_actions[:, self._action_index_pick_threshold]
```

Comment:
Same remark as above.

Reply:
Just copying reply from above:

```
mask_discrete = mask_invalid_actions[:, self._action_index_pick_threshold]
```
What this line does is creating a boolean mask of the elements (actions) in the batch that are discrete. It is done by taking the value (True or False) in the mask of invalid actions corresponding to the representative of the action "pick threshold", indicated by `self._action_index_pick_threshold`. If it is True, then the action "pick threshold" is invalid and therefore the action must be discrete and it will become `True` in `mask_discrete`. Does that make sense?